{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c740c32b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "numpy 1.21.2 \n",
      "pandas 1.3.4 \n",
      "statsmodels 0.13.2 \n",
      "sklearn 1.0.2\n",
      "ERROR! Session/line number was not unique in database. History logging moved to new session 42\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import numpy as np\n",
    "from numpy import loadtxt\n",
    "import pandas as pd\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "from sklearn import __version__ as sk__version__\n",
    "\n",
    "from sklearn.metrics import zero_one_loss\n",
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score\n",
    "from sklearn.metrics import classification_report, ConfusionMatrixDisplay\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.metrics import log_loss, roc_curve, roc_auc_score\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import tensorflow as tf\n",
    "from tensorflow import keras\n",
    "from tensorflow.keras import Sequential\n",
    "from tensorflow.keras.layers import Conv2D, Flatten, Dense\n",
    "from tensorflow.keras.optimizers import SGD\n",
    "\n",
    "\n",
    "print(\"\\nnumpy\", np.__version__,\n",
    "      \"\\npandas\", pd.__version__,\n",
    "      \"\\nstatsmodels\", sm.__version__,\n",
    "      \"\\nsklearn\", sk__version__)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ff3752e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reading the file and splitting our target column\n",
    "df = pd.read_csv (r'bank-additional.csv', sep = ';', engine= 'python')   \n",
    "# length = df.shape[0], 4119 x 21\n",
    "tussen = df[[\"y\"]]\n",
    "target = tussen.copy()\n",
    "df.drop(df.columns[len(df.columns)-1], axis=1, inplace=True)\n",
    "\n",
    "#Encoding all the catergorical variables and deleting them from the old dataframe\n",
    "categories = ['marital','education','default','housing','loan','contact','month','day_of_week','poutcome']\n",
    "df_encoded = pd.get_dummies(df['job'], prefix='job', drop_first=False)\n",
    "df_encoded.insert(0, 'ID', range(0, len(df_encoded)))\n",
    "del df['job']\n",
    "\n",
    "for x in categories:\n",
    "    df_encoded_tussen = pd.get_dummies(df[x], prefix=x, drop_first=True)\n",
    "    df_encoded_tussen.insert(0, 'ID', range(0, len(df_encoded_tussen)))\n",
    "    df_encoded = pd.merge(df_encoded, df_encoded_tussen, on = 'ID', how = 'outer')\n",
    "    del df[x]\n",
    "\n",
    "df.insert(0, 'ID', range(0, len(df)))\n",
    "features = pd.merge(df_encoded, df, on = 'ID', how = 'outer')\n",
    "y = pd.get_dummies(target, drop_first=True)\n",
    "del features['ID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7754a0ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  6.   , 148.   ,  72.   , ...,  33.6  ,   0.627,  50.   ],\n",
       "       [  1.   ,  85.   ,  66.   , ...,  26.6  ,   0.351,  31.   ],\n",
       "       [  8.   , 183.   ,  64.   , ...,  23.3  ,   0.672,  32.   ],\n",
       "       ...,\n",
       "       [  5.   , 121.   ,  72.   , ...,  26.2  ,   0.245,  30.   ],\n",
       "       [  1.   , 126.   ,  60.   , ...,  30.1  ,   0.349,  47.   ],\n",
       "       [  1.   ,  93.   ,  70.   , ...,  30.4  ,   0.315,  23.   ]])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR! Session/line number was not unique in database. History logging moved to new session 39\n"
     ]
    }
   ],
   "source": [
    "# TEST CODE\n",
    "# load the dataset\n",
    "dataset = loadtxt('pima-indians-diabetes.data.csv', delimiter=',')\n",
    "# split into input (X) and output (Y) variables\n",
    "X1 = dataset[:,0:8]\n",
    "y1 = dataset [:, 8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b88f3989",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR! Session/line number was not unique in database. History logging moved to new session 40\n",
      "Train on 4119 samples\n",
      "Epoch 1/100\n",
      "4119/4119 [==============================] - 0s 99us/sample - loss: 31214.6206 - acc: 0.8521\n",
      "Epoch 2/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.5755 - acc: 0.8905\n",
      "Epoch 3/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.5227 - acc: 0.8905\n",
      "Epoch 4/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.4841 - acc: 0.8905\n",
      "Epoch 5/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.4551 - acc: 0.8905\n",
      "Epoch 6/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.4330 - acc: 0.8905\n",
      "Epoch 7/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.4160 - acc: 0.8905\n",
      "Epoch 8/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.4028 - acc: 0.8905\n",
      "Epoch 9/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3925 - acc: 0.8905\n",
      "Epoch 10/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3843 - acc: 0.8905\n",
      "Epoch 11/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3778 - acc: 0.8905\n",
      "Epoch 12/100\n",
      "4119/4119 [==============================] - 0s 20us/sample - loss: 0.3724 - acc: 0.8905\n",
      "Epoch 13/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3681 - acc: 0.8905\n",
      "Epoch 14/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3645 - acc: 0.8905\n",
      "Epoch 15/100\n",
      "4119/4119 [==============================] - 0s 22us/sample - loss: 0.3615 - acc: 0.8905\n",
      "Epoch 16/100\n",
      "4119/4119 [==============================] - 0s 20us/sample - loss: 0.3591 - acc: 0.8905\n",
      "Epoch 17/100\n",
      "4119/4119 [==============================] - 0s 22us/sample - loss: 0.3570 - acc: 0.8905\n",
      "Epoch 18/100\n",
      "4119/4119 [==============================] - 0s 22us/sample - loss: 0.3554 - acc: 0.8905\n",
      "Epoch 19/100\n",
      "4119/4119 [==============================] - 0s 22us/sample - loss: 0.3539 - acc: 0.8905\n",
      "Epoch 20/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3527 - acc: 0.8905\n",
      "Epoch 21/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3517 - acc: 0.8905\n",
      "Epoch 22/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3508 - acc: 0.8905\n",
      "Epoch 23/100\n",
      "4119/4119 [==============================] - 0s 20us/sample - loss: 0.3501 - acc: 0.8905\n",
      "Epoch 24/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3495 - acc: 0.8905\n",
      "Epoch 25/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3489 - acc: 0.8905\n",
      "Epoch 26/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3484 - acc: 0.8905\n",
      "Epoch 27/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3480 - acc: 0.8905\n",
      "Epoch 28/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3477 - acc: 0.8905\n",
      "Epoch 29/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3474 - acc: 0.8905\n",
      "Epoch 30/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3472 - acc: 0.8905\n",
      "Epoch 31/100\n",
      "4119/4119 [==============================] - 0s 20us/sample - loss: 0.3469 - acc: 0.8905\n",
      "Epoch 32/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3467 - acc: 0.8905\n",
      "Epoch 33/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3466 - acc: 0.8905\n",
      "Epoch 34/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3465 - acc: 0.8905\n",
      "Epoch 35/100\n",
      "4119/4119 [==============================] - 0s 20us/sample - loss: 0.3463 - acc: 0.8905\n",
      "Epoch 36/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3462 - acc: 0.8905\n",
      "Epoch 37/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3461 - acc: 0.8905\n",
      "Epoch 38/100\n",
      "4119/4119 [==============================] - 0s 22us/sample - loss: 0.3460 - acc: 0.8905\n",
      "Epoch 39/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3460 - acc: 0.8905\n",
      "Epoch 40/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3459 - acc: 0.8905\n",
      "Epoch 41/100\n",
      "4119/4119 [==============================] - 0s 20us/sample - loss: 0.3458 - acc: 0.8905\n",
      "Epoch 42/100\n",
      "4119/4119 [==============================] - 0s 20us/sample - loss: 0.3458 - acc: 0.8905\n",
      "Epoch 43/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3457 - acc: 0.8905\n",
      "Epoch 44/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3457 - acc: 0.8905\n",
      "Epoch 45/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3457 - acc: 0.8905\n",
      "Epoch 46/100\n",
      "4119/4119 [==============================] - 0s 22us/sample - loss: 0.3457 - acc: 0.8905\n",
      "Epoch 47/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3456 - acc: 0.8905\n",
      "Epoch 48/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3456 - acc: 0.8905\n",
      "Epoch 49/100\n",
      "4119/4119 [==============================] - 0s 22us/sample - loss: 0.3456 - acc: 0.8905\n",
      "Epoch 50/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3456 - acc: 0.8905\n",
      "Epoch 51/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3456 - acc: 0.8905\n",
      "Epoch 52/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3456 - acc: 0.8905\n",
      "Epoch 53/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 54/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 55/100\n",
      "4119/4119 [==============================] - 0s 20us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 56/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 57/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 58/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 59/100\n",
      "4119/4119 [==============================] - 0s 20us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 60/100\n",
      "4119/4119 [==============================] - 0s 20us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 61/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 62/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 63/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 64/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 65/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 66/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 67/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 68/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 69/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 70/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 71/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 72/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 73/100\n",
      "4119/4119 [==============================] - 0s 20us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 74/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 75/100\n",
      "4119/4119 [==============================] - 0s 20us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 76/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 77/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 78/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 79/100\n",
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 80/100\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4119/4119 [==============================] - 0s 21us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 81/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 82/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 83/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 84/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 85/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 86/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 87/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 88/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 89/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 90/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 91/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 92/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 93/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 94/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 95/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 96/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 97/100\n",
      "4119/4119 [==============================] - 0s 17us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 98/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 99/100\n",
      "4119/4119 [==============================] - 0s 18us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Epoch 100/100\n",
      "4119/4119 [==============================] - 0s 19us/sample - loss: 0.3455 - acc: 0.8905\n",
      "Accuracy: 89.05\n"
     ]
    }
   ],
   "source": [
    "## NIEUWE NN CODE\n",
    "# Changing pandas dataframe to numpy array: input (X) and output (Y) variables\n",
    "X_new = features.values\n",
    "y_new = y.values # Hebben alleen de waarden 0?\n",
    "\n",
    "# Define the model\n",
    "model = Sequential()\n",
    "model.add(Dense(16, input_dim=54, activation='relu')) # First hidden layer: 16 nodes and uses the relu activation function\n",
    "model.add(Dense(12, activation='relu')) # Second hidden layer: 12 nodes and uses the relu activation function\n",
    "model.add(Dense(1, activation='sigmoid')) # Output layer: 1 node and uses the sigmoid activation function\n",
    "\n",
    "# Compile the model\n",
    "opt = SGD(learning_rate=0.01) # Gradient descent (with momentum) optimizer\n",
    "model.compile(loss = 'binary_crossentropy', optimizer = opt, metrics=['accuracy'])\n",
    "\n",
    "# Fit the model on the dataset\n",
    "model.fit(X_new, y_new, epochs=100, batch_size=64)\n",
    "\n",
    "# Evaluate the model\n",
    "_, accuracy = model.evaluate(X_new, y_new, verbose=0)\n",
    "print('Accuracy: %.2f' %(accuracy*100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bd1b06a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3457, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/ys/fbx59xk17y113h4831d2n8xr0000gn/T/ipykernel_16603/488755142.py\", line 25, in <module>\n",
      "    history = model.fit(train_features, train_targets, epochs=100, batch_size=64)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py\", line 727, in fit\n",
      "    use_multiprocessing=use_multiprocessing)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_arrays.py\", line 643, in fit\n",
      "    shuffle=shuffle)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py\", line 2537, in _standardize_user_data\n",
      "    y, self._feed_loss_fns, feed_output_shapes)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_utils.py\", line 741, in check_loss_and_target_compatibility\n",
      "    ' while using as loss `' + loss_name + '`. '\n",
      "ValueError: A target array with shape (3295, 2) was passed for an output of shape (None, 1) while using as loss `binary_crossentropy`. This loss expects targets to have the same shape as the output.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2077, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'ValueError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3457, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/ys/fbx59xk17y113h4831d2n8xr0000gn/T/ipykernel_16603/488755142.py\", line 25, in <module>\n",
      "    history = model.fit(train_features, train_targets, epochs=100, batch_size=64)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py\", line 727, in fit\n",
      "    use_multiprocessing=use_multiprocessing)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_arrays.py\", line 643, in fit\n",
      "    shuffle=shuffle)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py\", line 2537, in _standardize_user_data\n",
      "    y, self._feed_loss_fns, feed_output_shapes)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_utils.py\", line 741, in check_loss_and_target_compatibility\n",
      "    ' while using as loss `' + loss_name + '`. '\n",
      "ValueError: A target array with shape (3295, 2) was passed for an output of shape (None, 1) while using as loss `binary_crossentropy`. This loss expects targets to have the same shape as the output.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2077, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'ValueError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3377, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3474, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2080, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1125, in structured_traceback\n",
      "    tb_offset)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2077, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3457, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/ys/fbx59xk17y113h4831d2n8xr0000gn/T/ipykernel_16603/488755142.py\", line 25, in <module>\n",
      "    history = model.fit(train_features, train_targets, epochs=100, batch_size=64)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py\", line 727, in fit\n",
      "    use_multiprocessing=use_multiprocessing)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_arrays.py\", line 643, in fit\n",
      "    shuffle=shuffle)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py\", line 2537, in _standardize_user_data\n",
      "    y, self._feed_loss_fns, feed_output_shapes)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_utils.py\", line 741, in check_loss_and_target_compatibility\n",
      "    ' while using as loss `' + loss_name + '`. '\n",
      "ValueError: A target array with shape (3295, 2) was passed for an output of shape (None, 1) while using as loss `binary_crossentropy`. This loss expects targets to have the same shape as the output.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2077, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'ValueError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3377, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3474, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2080, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1125, in structured_traceback\n",
      "    tb_offset)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2077, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2960, in _run_cell\n",
      "    return runner(coro)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/async_helpers.py\", line 78, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3186, in run_cell_async\n",
      "    interactivity=interactivity, compiler=compiler, result=result)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3396, in run_ast_nodes\n",
      "    self.showtraceback()\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2080, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1143, in structured_traceback\n",
      "    chained_exceptions_tb_offset)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2077, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/jiunnni/opt/anaconda3/envs/seminar-marketing/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n"
     ]
    }
   ],
   "source": [
    "## OUDE NN CODE\n",
    "#Changing pandas dataframe to numpy array\n",
    "X = features.values\n",
    "y = y.values\n",
    "\n",
    "#Normalizing the data\n",
    "sc = StandardScaler()\n",
    "X = sc.fit_transform(X)\n",
    "\n",
    "ohe = OneHotEncoder()\n",
    "y = ohe.fit_transform(y).toarray()\n",
    "\n",
    "train_features, test_features, train_targets, test_targets = train_test_split(X, y, test_size=0.2,\n",
    "                                                                                  random_state=23)\n",
    "\n",
    "# Neural network\n",
    "model = Sequential()\n",
    "model.add(Dense(16, input_dim=54, activation='relu'))\n",
    "model.add(Dense(12, activation='relu'))\n",
    "model.add(Dense(1, activation='softmax'))\n",
    "\n",
    "opt = SGD(lr=0.01)\n",
    "model.compile(loss = \"binary_crossentropy\", optimizer = opt, metrics=['accuracy'])\n",
    "\n",
    "history = model.fit(train_features, train_targets, epochs=100, batch_size=64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa798e47",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "predictions = model.predict(test_features)\n",
    "\n",
    "\n",
    "# Use score method to get accuracy of model\n",
    "accuracy = metrics.accuracy_score(test_targets, predictions)\n",
    "print(\"Accuracy: \", + np.round(accuracy , 3))\n",
    "\n",
    "print(\"Confusion Matrix:\")\n",
    "print(confusion_matrix(test_targets, predictions))\n",
    "\n",
    "print(\"Classification Report\")\n",
    "print(classification_report(test_targets, predictions))\n",
    "    \n",
    "#Beginning the plotting of ROC-curve\n",
    "pred_prob = classifier.predict_proba(test_features)\n",
    "fpr, tpr, thresh = roc_curve(test_targets, pred_prob[:,1], pos_label=1)\n",
    "    \n",
    "#Plot roc curves\n",
    "plt.plot(fpr, tpr, linestyle='--',color='orange', label='SVM')\n",
    "\n",
    "# title\n",
    "plt.title('ROC curve')\n",
    "# x label\n",
    "plt.xlabel('False Positive Rate')\n",
    "# y label\n",
    "plt.ylabel('True Positive rate')\n",
    "\n",
    "plt.legend(loc='best')\n",
    "plt.savefig('ROC',dpi=300)\n",
    "plt.show();\n",
    "    \n",
    "#AUC Score\n",
    "auc_score = roc_auc_score(test_targets, pred_prob[:,1])\n",
    "print(\"AUC Score: \" + str(np.round(auc_score , 3)))\n",
    "    \n",
    "#Log-loss function\n",
    "print(\"Log-Loss: \" + str(np.round(log_loss(test_targets, predictions),3)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
